File: code/jvm/im-domain/src/main/kotlin/im/domain/wrappers/password/Password.kt
package im.domain.wrappers.password

import im.domain.Failure
import im.domain.Success

/**
 * Password wrapper class that enforces password validation rules.
 *
 * A password must not be blank and must be between 8 and 80 characters.
 */
data class Password(
    val value: String,
) {
    companion object {
        private val validator = PasswordValidator()
    }

    init {
        val validation = validator.validate(value)
        require(validation is Success) { (validation as Failure).value.toErrorMessage() }
    }

    override fun toString(): String = value
}

fun String.toPassword(): Password = Password(this)

File: code/client/web/src/Domain/wrappers/password/Password.ts
import { PasswordValidator } from './PasswordValidator';

export class Password {
    private static validator = new PasswordValidator();

    constructor(public readonly value: string) {
        const validation = Password.validator.validate(value);
        if (validation.length > 0) {
            throw new Error(validation.join('\n'));
        }
    }

    static fromString(value: string): Password {
        return new Password(value);
    }

    toString(): string {
        return this.value;
    }
}
